metric/producer/manager.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2019, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package producer
+
+import (
+	"sync"
+)
+
+type manager struct {
+	mu       sync.RWMutex
+	producers []Producer
```

Comment:
Consider using a `map[string]Producer`.

Reply:
i had considered that before but map cannot be unmodifiable. However slices are unmodifiable implicitly.

Reply:
You can implement your own immutable map/slice (also not sure that slices are immutable) by cloning the current map/slice and apply the mutation operation on the clone then swap the pointers to the map/slice.